### PR TITLE
fw_info: Add fw_info_ext_api_provide() as an EXT_API.

### PR DESCRIPTION
**Pull request by oyvindronningstad**
Allows MCUboot to provide EXT_APIs to its subimages.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

MCUboot PR: https://github.com/NordicPlayground/fw-nrfconnect-mcuboot/pull/84

Explanation:
When B0 boots an image, it first checks whether to give it access to some functions in it own code. This is what is called an EXT_API. When it finds a request from the child image, it tries to fulfill it by seeing what APIs are provided by itself, or - maybe in the future - other images. It does this with a function called fw_info_ext_api_provide(). The problem arises when e.g. MCUboot is the child image, and itself boots a child image. B0 is not able to fulfill EXT_API requests from this "grand-child" image because a) it's not aware that it exists, and b) EXT_APIs are shared through RAM, so it has to happen immediately before booting the image.

This PR provides the ext_api_provide() function as an EXT_API so that MCUboot can do B0's job of fulfilling the app's EXT_API requests before booting it.

### CODE REVIEW

**Comment by oyvindronningstad** [comment]
DNM until https://github.com/NordicPlayground/fw-nrfconnect-mcuboot/pull/84 is merged

File: subsys/fw_info/fw_info.c [hunk 1]
```diff
@@ -101,14 +105,19 @@ static const struct fw_info_ext_api_request *skip_ext_apis(
 }
 
 
+#ifndef CONFIG_EXT_API_PROVIDE_EXT_API_REQUIRED
```
Comment (hakonfam) [inline subsys/fw_info/fw_info.c]: Could you explain why we have this logic? To me it is not clear why we have these two variants

File: subsys/fw_info/fw_info.c [hunk 2]
```diff
@@ -66,14 +66,17 @@ static bool ext_api_satisfies_req(const struct fw_info_ext_api * const ext_api,
 		&& ((ext_api->ext_api_flags & req_flags) == req_flags));
 }
 
+#ifndef CONFIG_EXT_API_PROVIDE_EXT_API_REQUIRED
```
Comment (hakonfam) [inline subsys/fw_info/fw_info.c]: Could you explain why this check is here? The way the EXT_API's work are kinda magic, so IMO we should document as much as we can in the code to make it easier to understand.

**Comment by oyvindronningstad** [comment]
> I don't understand this feature enough to approve, so a bit of education is needed :)

See description :).

File: subsys/fw_info/fw_info.c (hunk 1 above)
Comment (oyvindronningstad) [reply subsys/fw_info/fw_info.c]: Inside this `#ifndef` is the proper implementation of `fw_info_ext_api_provide()`. If `CONFIG_EXT_API_PROVIDE_EXT_API_REQUIRED` is `y`, we want to call the EXT_API instead, which happens inside the `#else`.

I'll add this in a comment in the code as well :).

Comment (oyvindronningstad) [reply subsys/fw_info/fw_info.c]: Added

**Comment by hakonfam** [comment]
> > I don't understand this feature enough to approve, so a bit of education is needed :)
> 
> See description :).

Would it make sense to include that info in the commit message, or the code somehow?

File: subsys/fw_info/fw_info.c (hunk 1 above)
Comment (hakonfam) [reply subsys/fw_info/fw_info.c]: Just for my own understanding, would mcuboot still be capable of exposing its own EXT_API?

Comment (oyvindronningstad) [reply subsys/fw_info/fw_info.c]: Not really as it is right now.

Comment (oyvindronningstad) [reply subsys/fw_info/fw_info.c]: > > > I don't understand this feature enough to approve, so a bit of education is needed :)
> > 
> > 
> > See description :).
> 
> Would it make sense to include that info in the commit message, or the code somehow?

Good point, added to commit message.

**Comment by oyvindronningstad** [comment]
> > > I don't understand this feature enough to approve, so a bit of education is needed :)
> > 
> > 
> > See description :).
> 
> Would it make sense to include that info in the commit message, or the code somehow?

Good point, added to commit message.

File: subsys/fw_info/fw_info.c [hunk 3]
```diff
@@ -90,26 +106,22 @@ static const struct fw_info_ext_api *find_ext_api(
 }
 
 
-static const struct fw_info_ext_api_request *skip_ext_apis(
-					const struct fw_info * const fw_info)
+/**This is the "proper" implementation of fw_info_ext_api_provide(). If
```
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/fw_info/fw_info.c]: Fixed

File: subsys/fw_info/fw_info.c (hunk 1 above)
Comment (oyvindronningstad) [reply subsys/fw_info/fw_info.c]: Should I add the possibility for MCUboot to share its own APIs? Or should we do it later?

Comment (hakonfam) [reply subsys/fw_info/fw_info.c]: Lets wait with that.

**Pull request merged**